### PR TITLE
[dtensor] fix dtype/device conversion on nn.Modules

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111162

This PR fixes the dtype conversion issue when calling nn.Module.to after
swapping parameters with DTensors, the problem was nn.Module._apply by
default uses a `.data` setting way to swap out storages directly instead
of just replacing the nn.Parameters. For wrapper tensor subclasses, it
would not work as wrapper tensor subclass is different from normal
tensor as its real storage points separately, i.e.

* DTensor wrapper subclass instance `storage` does not point to real storage
* The real storage of DTensor pointed to its local shard. But if we do `.data=` it seems that the change will not propagate into the DTensor's local shard

This probably also fix https://github.com/pytorch/pytorch/issues/102812

### CODE REVIEW

**Comment by awgu** [comment]
I wonder if this means that this code in FSDP to move parameters from CPU to GPU is broken for DTensor:
https://github.com/pytorch/pytorch/blob/8f90be447800a5c624d96b928387cb6f21509d37/torch/distributed/fsdp/_init_utils.py#L993-L995

**Comment by wanchaol** [comment]
> I wonder if this means that this code in FSDP to move parameters from CPU to GPU is broken for DTensor:
> 
> https://github.com/pytorch/pytorch/blob/8f90be447800a5c624d96b928387cb6f21509d37/torch/distributed/fsdp/_init_utils.py#L993-L995

if the `params` and `buffers` are DTensors, possibly yeah, the wrapper class will move to GPU, but the local_shard likely not. Do we currently have DTensors use this API for state_dict?

I'll revise the test to add device conversion too.

**Comment by wanchaol** [comment]
@albanD @ezyang wondering if we have a way to only turn on this flag for wrapper based tensor subclasses? I think all wrapper subclasses will hit the same issue, but if I turn this flag on in dtensor package init, CI will fail because some dtensor tests run first and it never set it back (which triggers BC for other tests)

**Comment by ezyang** [comment]
as i said in chat, you probably should just add some logic to nn module itself to do something different when you have subclasses

**Comment by albanD** [comment]
The reason this setting was never turned on is that is it a pretty big semantic change and will break quite a bit of user code in the wild. If an optimizer was holding to the original Tensor, it will now hold a different Tensor from the one that is in the Module (and gets gradients).

Something similar to `.set_()` could be used to preserve the Tensor identity but it doesn't allow cross-device/dtype (for good reasons).
I think the two options are:
- Use assignment as proposed here and deal with the fact that it is a semantic change. Maybe even try and figure out what we could do to enable that future by default going forward.
- Introduce a new native op `base.move_(src)` that will perform a full move (or swap if that is simpler?) of the Tensor into the other (we can make it have slightly saner semantic than .data= by doing a full move).

**Comment by wanchaol** [comment]
> * Introduce a new native op `base.move_(src)` that will perform a full move (or swap if that is simpler?) of the Tensor into the other (we can make it have slightly saner semantic than .data= by doing a full move).

Sounds good, thanks for the suggestions! @albanD @ezyang !! I'm interested in this which looks pretty clean semantic to me and we can handle subclassing inside this op too. I'll go with suggestions from @ezyang consider that this is simpler and I'll add a TODO there to explore the `move_` option

File: torch/distributed/_tensor/__init__.py [hunk 1]
```diff
@@ -360,3 +360,11 @@ def zeros(
...
 if not torch._running_with_deploy():
     import torch.distributed._tensor._dynamo_utils
+
+
+# disable param.data set when registering nn.Parameters, and use
+# the future/new behavior (overwrite module parameters) when using
+# the DTensor package
+# NOTE: wrapper tensor subclass storage is different from the original
+# tensor of nn.Parameters, so we should avoid using the `.data` path
+torch.__future__.set_overwrite_module_params_on_conversion(True)
```
Comment (fduwjj) [inline torch/distributed/_tensor/__init__.py:370]: Will this affect TP + FSDP integration?

Comment (wanchaol) [reply torch/distributed/_tensor/__init__.py:370]: I'll remove this and modify nn.module instead

**Comment by ezyang** [comment]
@albanD what is a "full move"? This is not a thing lol, you often literally do not have enough space in C++ to rep the extra data

**Comment by wanchaol** [comment]
> @albanD what is a "full move"? This is not a thing lol, you often literally do not have enough space in C++ to rep the extra data

@ezyang I think what @albanD referred to is more like a `std::move` to apply to the tensor? `tensor.move_` to me looks like we still preserve this current `tensor` identity (i.e. memory address of this tensor or wrapper tensor), but under the hood we swap the tensor storage ptr to the new one?

For now I implemented the traceable tensor subclass check in module.py directly, lmk if we should create a issue to track this `tensor.move_` discussion :)

**Comment by ezyang** [comment]
@albanD is talking about something like this


```
PyObject* THPModule_swap(PyObject* _unused, PyObject* args) {
  PyObject* a_ = nullptr;
  PyObject* b_ = nullptr;
  if (!PyArg_ParseTuple(args, "OO", &a_, &b_)) {
    return nullptr;
  }

  TORCH_CHECK(THPVariable_Check(a_));
  TORCH_CHECK(THPVariable_Check(b_));
  // TODO: Add a check to make sure the user didn't subclass
  // THPVariable from C and added new fields.
  // Not sure how but that should be rare

  // Make sure we use managed dict otherwise we would read out of bound!
  TORCH_CHECK(PyType_HasFeature(Py_TYPE(a_), Py_TPFLAGS_MANAGED_DICT));

  // A bunch of stuff is before the pointer: the gc prev/next, __dict__
  // and weakref list. So we offset back by 4
  // PyObject should be PyDictOrValues but we don't have access to it
  void* a = (void*)a_ - 4*sizeof(PyObject*);
  void* b = (void*)b_ - 4*sizeof(PyObject*);
  auto actual_size = sizeof(THPVariable) + 4*sizeof(PyObject*);

  // Swap the full content of the PyObjects
  void* tmp = malloc(actual_size);
  memcpy(tmp, a, actual_size);
  memcpy(a, b, actual_size);
  memcpy(b, tmp, actual_size);
  free(tmp);

  // The only thing that we must preserve is the gc stuff!
  // To make sure that everyone holding a PyObject* reference
  // to this is still correct.
  Py_ssize_t tmp_refcnt = Py_REFCNT(a);
  Py_SET_REFCNT(a, Py_REFCNT(b));
  Py_SET_REFCNT(b, tmp_refcnt);

  PyGC_Head* a_gc = _Py_AS_GC(a_);
  PyGC_Head* b_gc = _Py_AS_GC(b_);
  uintptr_t tmp_gc_head;
  tmp_gc_head = a_gc->_gc_next;
  a_gc->_gc_next = b_gc->_gc_next;
  b_gc->_gc_next = tmp_gc_head;
  tmp_gc_head = a_gc->_gc_prev;
  a_gc->_gc_prev = b_gc->_gc_prev;
  b_gc->_gc_prev = tmp_gc_head;

  Py_RETURN_NONE;
}
```

**Review by albanD — CHANGES_REQUESTED** [review]
This changes the behavior too much I think. Let me put up a proposal for a swap.
